KOGITO 2707 - add trusty image

r00ta opened the pull request:
Jira: https://issues.redhat.com/browse/KOGITO-2707

This PR aims to add the trusty image. The trusty service is very similar to the data-index service in terms of dependencies (they both depend on infinispan and kafka). You can find the trusty service implementation here https://github.com/kiegroup/kogito-apps/tree/master/trusty . 
This image will be used in the operator for the integration with the trusty initiative. 

Many thanks for submitting your Pull Request :heart:! 

Please make sure your PR meets the following requirements:

- [x] You have read the [contributors guide](README.md#contributing-to-kogito-images-repository)
- [x] Pull Request title is properly formatted: `[KOGITO-XYZ] Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket
- [x] Your feature/bug fix has a testcase that verifies it
- [ ] You've tested the new feature/bug fix in an actual OpenShift cluster

ricardozanini commented on kogito-imagestream.yaml:
```diff
@@ -132,6 +132,27 @@ items:
+        openshift.io/provider-display-name: Kie Group.
+    spec:
+      tags:
+        - name: '0.12.0-rc1'
```
0.12 will be released today, after the cut we will need to rebase and adjust.

r00ta commented on README.md:
````diff
@@ -594,6 +596,30 @@ The [Kogito Operator](https://github.com/kiegroup/kogito-cloud-operator) can be
+
+Basic usage
+```bash
+$ docker run -it --env QUARKUS_INFINISPAN_CLIENT_SERVER_LIST=my-infinispan-server:11222 quay.io/kiegroup/kogito-trusty:latest
````
I might be wrong, but currently this is not implemented in the data index image: the kafka topics/servers are injected with the operator. @radtriste @ricardozanini please correct me if I'm wrong : )
Personally, It would make sense to me to have the possibility to set the kafka servers/topics also in the image with enviroment variables (if somebody is using the images without the operator). If you agree, I would open another PR and implement it for the data index and the trusty images.

radtriste replied:
This is all a problem in the documentation but Kafka env var are already taken account...
You can already set the env variables for Kafka here in the example. Those env variables are then used by the jar artifact. Images have no specificities for those env variables.
It is written in image.yaml description just as a documentation and indeed, Kafka vars are missing, but I am not sure we should write them here in fact ... it could be deleted.
@ricardozanini you confirm ?

r00ta replied:
Hi @radtriste , I've just tried but apparently the variable `KAFKA_BOOTSTRAP_SERVERS` is ignored by the data-index. Here the test that I've done: 
1) In one shell run `docker run -it jboss/infinispan-server` -> it prints the credentials
2) In another shell run `docker run -it --link inf --env QUARKUS_INFINISPAN_CLIENT_SERVER_LIST=inf:11222 --env INFINISPAN_USERNAME=jp5cDuo2Ee --env INFINISPAN_PASSWORD=GjLqEmjNAl --env KAFKA_BOOTSTRAP_SERVERS=mykafka:9999 quay.io/kiegroup/kogito-data-index:latest`

In the debug messages I see that it is trying to connect to `localhost:9092` instead of `mykafka:9999`
```
[Consumer clientId=consumer-kogito-data-index-usertaskinstances-4, groupId=kogito-data-index-usertaskinstances] Connection to node -1 (localhost/127.0.0.1:9092) could not be established. Broker may not be available.
```
![Screenshot from 2020-07-10 09-52-45](https://user-images.githubusercontent.com/18282531/87130516-51686a80-c293-11ea-9094-3ea5fc5c1ddf.png)

radtriste replied:
Could be related to https://github.com/quarkusio/quarkus/issues/7791#issuecomment-656504672. apps are still using Quarkus 1.4.1 ... https://github.com/kiegroup/kogito-apps/blob/e0e7374385db5312b65bfa03f026ad7496eb7614/pom.xml#L19

radtriste replied:
do you have any default configuration of `mp.messaging[...]` into your `src/main/resources/application.properties` file ?

r00ta replied:
I've tried with the latest data index image, I'm checking what do we have in that version

r00ta replied:
Yes, in the data index the bootstrap servers were and are currently set https://github.com/kiegroup/kogito-apps/blob/e0e7374385db5312b65bfa03f026ad7496eb7614/data-index/data-index-service/src/main/resources/application.properties#L63

r00ta replied:
I'll try to recompile everything with the current version and check it

radtriste replied:
with Quarkus 1.4.1, you will need to set env vars:
- `MP_MESSAGING_INCOMING_KOGITO_PROCESS_INSTANCES_EVENTS_BOOTSTRAP_SERVERS`
- `MP_MESSAGING_INCOMING_KOGITO_PROCESS_DOMAIN_EVENTS_BOOTSTRAP_SERVERS`
- `MP_MESSAGING_INCOMING_KOGITO_USERTASKINSTANCES_EVENTS_BOOTSTRAP_SERVERS`
- `MP_MESSAGING_INCOMING_KOGITO_USERTASKDOMAIN_EVENTS_BOOTSTRAP_SERVERS`
if I remember correctly

r00ta replied:
wops, I've just noticed that the kogito-apps are still using quarkus 1.4.1, while kogito-runtimes has the latest 1.6.0. I'm trying to upgrade it but there are some failures here and there.. I'll ask if there is already a plan for it

r00ta replied:
Update: the kogito-apps repo is being upgraded to quarkus 1.6.0 with the PR https://github.com/kiegroup/kogito-apps/pull/319 . I've tested it and the injection of the env variable KAFKA_BOOTSTRAP_SERVERS works as expected with that version.

spolti commented on kogito-trusty-overrides.yaml:
```diff
@@ -0,0 +1,46 @@
+  - name: "HTTP_PORT"
+    example: "9090"
+    description: "Defines which port the Trusty Container Image will listen on."
+  - name: "QUARKUS_INFINISPAN_CLIENT_SERVER_LIST"
```
is the trusty-service suppose to work only with Quarkus?
What about springboot?

danielezonca replied:
This service is like other Kogito-apps services so it is immutable and quarkus only. User is not supposed to change it but only use it (like data-index)

r00ta commented:
Hi @ricardozanini , @radtriste I see this jira https://issues.redhat.com/browse/KOGITO-2046 and the related pull request. I propose that as soon as it is merged, I'll rebase this pr and do the same for this image

ricardozanini commented:
Depends on #189

spolti commented:
can be merged after https://github.com/kiegroup/kogito-images/pull/191

jiripetrlik commented on kogito-imagestream.yaml:
```diff
@@ -132,6 +132,27 @@ items:
           from:
             kind: DockerImage
             name: quay.io/kiegroup/kogito-data-index:0.12.0-rc1
+  - kind: ImageStream
```
@r00ta @ricardozanini 
Do we have any specific kogito-trusty image? Can you point me to some source code or Dockerfile about how it is build and what exactly is inside?

ricardozanini replied:
This PR is supposed to build the trusty image :)

You have to use cekit to build it, we do not create Dockerfiles manually: https://docs.cekit.io/en/latest/

r00ta commented:
Hi, I've removed the infinispan dependencies to align with  https://github.com/kiegroup/kogito-images/pull/189

r00ta commented:
jenkins retest this please

r00ta commented:
jenkins retest this please

r00ta commented:
Hi @ricardozanini @radtriste @spolti , can we move forward with this?

ricardozanini commented:
@danielezonca @jiripetrlik @MarianMacik are you guys ok with this? We can release this image as part of 0.13.0 if we merge it up until Friday. I'd say to wait @spolti come back on Monday, but if we have something to update, we can follow up in the next version.

radtriste commented:
jenkins retest this please

radtriste commented:
jenkins retest this please

danielezonca commented:
> @danielezonca @jiripetrlik @MarianMacik are you guys ok with this? We can release this image as part of 0.13.0 if we merge it up until Friday. I'd say to wait @spolti come back on Monday, but if we have something to update, we can follow up in the next version.

To be honest it is not mandatory to merge this today. I have approved the PR and it is ready from my point of view but if others want other changes it is not a problem for me :)

r00ta commented:
> as for explainability servive, please check what @vaibhavjainwiz has changed in his PR and update accordingly

Hi @radtriste, thanks! I've removed the infinispan explicit env variables and I've added a test to check that the service starts properly. It should be aligned now

ricardozanini commented:
Will merge as soon as the CI finish successfully.

radtriste commented on tests/features/kogito-trusty.feature:
```diff
@@ -0,0 +1,49 @@
+      | SCRIPT_DEBUG                  | true            |
+      | ENABLE_PERSISTENCE            | true            |
+      | QUARKUS_INFINISPAN_CLIENT_SERVER_LIST     | 172.18.0.1:11222  |
+      | quarkus.infinispan-client.use-auth        | true              |
```
```suggestion
      | QUARKUS_INFINISPAN_CLIENT_USE_AUTH        | true              |
```

radtriste commented on tests/features/kogito-trusty.feature:
```diff
@@ -0,0 +1,49 @@
+      | quarkus.infinispan-client.use-auth        | true              |
+      | QUARKUS_INFINISPAN_CLIENT_AUTH_USERNAME   | IamNotExist       |
+      | QUARKUS_INFINISPAN_CLIENT_AUTH_PASSWORD   | hard2guess        |
+      | quarkus.infinispan-client.auth-realm      | SecretRealm       |
```
```suggestion
      | QUARKUS_INFINISPAN_CLIENT_AUTH_REALM       | SecretRealm       |
```

radtriste commented on tests/features/kogito-trusty.feature:
```diff
@@ -0,0 +1,49 @@
+      | QUARKUS_INFINISPAN_CLIENT_AUTH_USERNAME   | IamNotExist       |
+      | QUARKUS_INFINISPAN_CLIENT_AUTH_PASSWORD   | hard2guess        |
+      | quarkus.infinispan-client.auth-realm      | SecretRealm       |
+      | quarkus.infinispan-client.sasl-mechanism  | COOLGSSAPI        |
```
```suggestion
      | QUARKUS_INFINISPAN_CLIENT_SASL_MECHANISM  | COOLGSSAPI        |
```

radtriste commented on tests/features/kogito-trusty.feature:
```diff
@@ -0,0 +1,49 @@
+      | quarkus.infinispan-client.sasl-mechanism  | COOLGSSAPI        |
+    Then container log should contain kogito.jobs-service.persistence=infinispan
+    Then container log should contain QUARKUS_INFINISPAN_CLIENT_SERVER_LIST=172.18.0.1:11222
+    Then container log should contain quarkus.infinispan-client.use-auth=true
```
```suggestion
    Then container log should contain QUARKUS_INFINISPAN_CLIENT_USE_AUTH=true
```

radtriste commented on tests/features/kogito-trusty.feature:
```diff
@@ -0,0 +1,49 @@
+    Then container log should contain quarkus.infinispan-client.use-auth=true
+    And container log should contain QUARKUS_INFINISPAN_CLIENT_AUTH_PASSWORD=hard2guess
+    And container log should contain QUARKUS_INFINISPAN_CLIENT_AUTH_USERNAME=IamNotExist
+    And container log should contain quarkus.infinispan-client.auth-realm=SecretReal
```
```suggestion
    And container log should contain QUARKUS_INFINISPAN_CLIENT_AUTH_REALM=SecretReal
```

radtriste commented on tests/features/kogito-trusty.feature:
```diff
@@ -0,0 +1,49 @@
+    And container log should contain QUARKUS_INFINISPAN_CLIENT_AUTH_PASSWORD=hard2guess
+    And container log should contain QUARKUS_INFINISPAN_CLIENT_AUTH_USERNAME=IamNotExist
+    And container log should contain quarkus.infinispan-client.auth-realm=SecretReal
+    And container log should contain quarkus.infinispan-client.sasl-mechanism=COOLGSSAPI
```
```suggestion
    And container log should contain QUARKUS_INFINISPAN_CLIENT_SASL_MECHANISM=COOLGSSAPI
```

radtriste commented on modules/kogito-trusty/added/kogito-app-launch.sh:
```diff
@@ -0,0 +1,25 @@
+)
+source ${KOGITO_HOME}/launch/configure.sh
+#############################################
+
```
You will need to `printenv` for the test:
https://github.com/kiegroup/kogito-images/pull/189/files#diff-500da60eb56e0f8c99d8254fb11183a8R22

r00ta replied:
Yeah you're right!

radtriste commented:
@r00ta need a rebase as well :)

r00ta commented:
jenkins retest this please

r00ta commented:
Hi @radtriste @ricardozanini , the build looks good. Are we ok to merge or we wait until Monday? No hurry from my side, but it would be nice to have it in the next release 0.13
